### PR TITLE
v16: Expose the unstable metadata v16

### PR DESCRIPTION
**Pull request by lexnv**
This PR exposes the *unstable* metadata V16. The metadata is exposed under the unstable u32::MAX number.

Developers can start experimenting with the new features of the metadata v16. *Please note that this metadata is under development and expect breaking changes until stabilization.*

The `ExtrinsicMetadata` trait receives a breaking change. Its associated type `VERSION` is rename to `VERSIONS` and now supports a constant static list of metadata versions.
The versions implemented for `UncheckedExtrinsic` are v4 (legacy version) and v5 (new version).

For metadata collection, it is assumed that all `TransactionExtensions` are under version 0.

Builds on top of: https://github.com/paritytech/polkadot-sdk/pull/5274

Closes: https://github.com/paritytech/polkadot-sdk/issues/5980
Closes: https://github.com/paritytech/polkadot-sdk/issues/5347
Closes: https://github.com/paritytech/polkadot-sdk/issues/5285

cc @paritytech/subxt-team

### CODE REVIEW

**Review by pkhry — APPROVED** [review]

File: substrate/frame/metadata-hash-extension/Cargo.toml [hunk 1]
```diff
@@ -25,7 +25,7 @@ substrate-test-runtime-client = { workspace = true }
 sp-api = { workspace = true, default-features = true }
 sp-transaction-pool = { workspace = true, default-features = true }
 merkleized-metadata = { workspace = true }
-frame-metadata = { features = ["current"], workspace = true, default-features = true }
+frame-metadata = { features = ["current","unstable"], workspace = true, default-features = true }
```
Comment (niklasad1) [inline substrate/frame/metadata-hash-extension/Cargo.toml:28]: I  think  `"unstable"`  will activate both "current" and "unstable IIRC

```suggestion
frame-metadata = { features = ["current", "unstable"], workspace = true, default-features = true }
```

File: prdoc/pr_5732.prdoc [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      - Runtime Dev
+      - Runtime User
+    description: |
+      This PR exposes the *unstable* metadata V16. The metadata is exposed under the unstable u32::MAX number.
+      Developers can start experimenting with the new features of the metadata v16. Please note that
+      this metadata is under development and expect breaking changes until stabilization.
+
+crates:
+  - name: sp-metadata-ir
+    bump: patch
```
Comment (niklasad1) [inline prdoc/pr_5732.prdoc:17]: Shouldn't this be minor i.e, you add new functionality in backward compatible manner?

File: substrate/primitives/metadata-ir/src/v14.rs [hunk 3]
```diff
@@ -151,7 +151,7 @@ impl From<ExtrinsicMetadataIR> for ExtrinsicMetadata {
 	fn from(ir: ExtrinsicMetadataIR) -> Self {
 		ExtrinsicMetadata {
 			ty: ir.ty,
-			version: ir.version,
+			version: *ir.versions.get(0).expect("Metadata V14 supports only one version"),
```
Comment (niklasad1) [inline substrate/primitives/metadata-ir/src/v14.rs:154]: Is there any plan to support several versions here and how to to find the latest version? I guess it will be last item in the Vec?

So my question is really if you need to do a linear search to find the specific version?

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs [hunk 4]
```diff
@@ -117,7 +117,7 @@ pub fn expand_runtime_metadata(
 					pallets: #scrate::__private::vec![ #(#pallets),* ],
 					extrinsic: #scrate::__private::metadata_ir::ExtrinsicMetadataIR {
 						ty,
-						version: <#extrinsic as #scrate::sp_runtime::traits::ExtrinsicMetadata>::VERSION,
+						versions: #scrate::__private::vec![ <#extrinsic as #scrate::sp_runtime::traits::ExtrinsicMetadata>::VERSION ],
```
Comment (jsdw) [inline substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs:120]: Maybe similar to what Niklas asked but, I guess this should now contain 2 versions, 4 and 5, as of https://github.com/paritytech/polkadot-sdk/pull/3685 merging?

File: substrate/primitives/metadata-ir/src/unstable.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+impl From<ExtrinsicMetadataIR> for ExtrinsicMetadata {
+	fn from(ir: ExtrinsicMetadataIR) -> Self {
+		ExtrinsicMetadata {
+			versions: ir.versions.into_iter().map(Into::into).collect(),
+			address_ty: ir.address_ty,
+			call_ty: ir.call_ty,
+			signature_ty: ir.signature_ty,
+			extra_ty: ir.extra_ty,
```
Comment (jsdw) [inline substrate/primitives/metadata-ir/src/unstable.rs:174]: Just a reminder that we should remove this type from V16; can't remember if there is an issue for this already :)

File: substrate/primitives/metadata-ir/src/unstable.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+		}
+	}
+}
+
+impl From<ExtrinsicMetadataIR> for ExtrinsicMetadata {
+	fn from(ir: ExtrinsicMetadataIR) -> Self {
+		ExtrinsicMetadata {
+			versions: ir.versions.into_iter().map(Into::into).collect(),
+			address_ty: ir.address_ty,
+			call_ty: ir.call_ty,
```
Comment (jsdw) [inline substrate/primitives/metadata-ir/src/unstable.rs:172]: Is this the same as `OuterEnums.call_enum_ty`? if so, can remove this also!

File: substrate/primitives/metadata-ir/src/types.rs [hunk 7]
```diff
@@ -170,8 +170,8 @@ pub struct ExtrinsicMetadataIR<T: Form = MetaForm> {
 	///
 	/// Note: Field used for metadata V14 only.
 	pub ty: T::Type,
-	/// Extrinsic version.
-	pub version: u8,
```
Comment (jsdw) [inline substrate/primitives/metadata-ir/src/types.rs:174]: Reminder that `extensions` below will also need to have a shape more like `Vec<(extensions_version, Vex<TransactionExtensionMetadataIR>)>` or something, but as with the other things I assume this will be a separate PR :)

**Review by jsdw — APPROVED** [review]
LGTM!

I saw a couple of things which I assume will be followup PRs, and I suppose this is mainly just about making V16 exposed, but added notes anyways just in case :)

File: substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs (hunk 4 above)
Comment (gui1117) [reply substrate/frame/support/procedural/src/construct_runtime/expand/metadata.rs:120]: `ExtrinsicMetadata::VERSION` should be a vec itself (or static slice).

File: substrate/primitives/metadata-ir/src/unstable.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+impl From<MetadataIR> for RuntimeMetadataV16 {
+	fn from(ir: MetadataIR) -> Self {
+		RuntimeMetadataV16::new(
+			ir.pallets.into_iter().map(Into::into).collect(),
+			ir.extrinsic.into(),
+			ir.apis.into_iter().map(Into::into).collect(),
+			ir.outer_enums.into(),
+			// Substrate does not collect yet the custom metadata fields.
+			// This allows us to extend the V15 easily.
```
Comment (gui1117) [inline substrate/primitives/metadata-ir/src/unstable.rs:47]:
```suggestion
			// This allows us to extend the V16 easily.
```

am I wrong?

File: substrate/primitives/metadata-ir/src/unstable.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+		}
+	}
+}
+
+impl From<TransactionExtensionMetadataIR> for TransactionExtensionMetadata {
+	fn from(ir: TransactionExtensionMetadataIR) -> Self {
+		TransactionExtensionMetadata {
+			identifier: ir.identifier,
+			ty: ir.ty,
+			additional_signed: ir.implicit,
```
Comment (gui1117) [inline substrate/primitives/metadata-ir/src/unstable.rs:162]: maybe metadata v16 should also rename it to `implicit`?

File: substrate/primitives/metadata-ir/src/v14.rs (hunk 3 above)
Comment (jsdw) [reply substrate/primitives/metadata-ir/src/v14.rs:154]: Agreed that V14 should continue to return the lowest version supported.

File: substrate/primitives/metadata-ir/src/unstable.rs [hunk 10]
```diff
@@ -156,22 +156,24 @@ impl From<PalletConstantMetadataIR> for PalletConstantMetadata {
...
+		TransactionExtensionMetadata { identifier: ir.identifier, ty: ir.ty, implicit: ir.implicit }
 	}
 }
 
 impl From<ExtrinsicMetadataIR> for ExtrinsicMetadata {
 	fn from(ir: ExtrinsicMetadataIR) -> Self {
+		// Presume all extensions work for all versions.
+		// This is true for now.
+		let transaction_extensions_by_version = ir
+			.versions
```
Comment (jsdw) [inline substrate/primitives/metadata-ir/src/unstable.rs:168]: That's a different version, isn't it? `ir.versions` is the transaction version (4 or 5 nowadays), but `transaction)extensions_by_version` maps the _transaction extension version_ to the list of extensions that apply. 

Thus, at the moment we'll have just one version, 0u8, which maps to the list of transaction extensions we have. I guess the `ExtrinsicMetadataIR` should be tweaked to record this information though, so that it's just mapped more 1-to-1 into this ExtrinsicMetadata?

File: substrate/primitives/runtime/src/traits/mod.rs [hunk 11]
```diff
@@ -1409,10 +1409,10 @@ impl SignaturePayload for () {
 
 /// Implementor is an [`Extrinsic`] and provides metadata about this extrinsic.
 pub trait ExtrinsicMetadata {
-	/// The format version of the `Extrinsic`.
+	/// The format versions of the `Extrinsic`.
 	///
 	/// By format is meant the encoded representation of the `Extrinsic`.
```
Comment (jsdw) [inline substrate/primitives/runtime/src/traits/mod.rs:1414]:
```suggestion
	/// By format we mean the encoded representation of the `Extrinsic`.
```

File: Cargo.toml [hunk 12]
```diff
@@ -846,7 +846,7 @@ macro_magic = { version = "0.5.1" }
 maplit = { version = "1.0.2" }
 memmap2 = { version = "0.9.3" }
 memory-db = { version = "0.32.0", default-features = false }
-merkleized-metadata = { version = "0.1.0" }
+merkleized-metadata = { git = "https://github.com/lexnv/merkleized-metadata.git", branch = "lexnv/update-frame-md" }
```
Comment (bkchr) [inline Cargo.toml:849]:
```suggestion
merkleized-metadata = "0.1.1"
```

File: substrate/primitives/metadata-ir/src/v15.rs [hunk 13]
```diff
@@ -100,7 +100,7 @@ impl From<TransactionExtensionMetadataIR> for SignedExtensionMetadata {
 impl From<ExtrinsicMetadataIR> for ExtrinsicMetadata {
 	fn from(ir: ExtrinsicMetadataIR) -> Self {
 		ExtrinsicMetadata {
-			version: ir.version,
+			version: *ir.versions.get(0).expect("Metadata V15 supports only one version"),
```
Comment (bkchr) [inline substrate/primitives/metadata-ir/src/v15.rs:103]:
```suggestion
			version: *ir.versions.iter().min().expect("Metadata V15 supports only one version"),
```

File: substrate/primitives/metadata-ir/src/unstable.rs (hunk 10 above)
Comment (bkchr) [reply substrate/primitives/metadata-ir/src/unstable.rs:168]: Yeah the `extension_version` is a different version. Here we should for now just assume `0`.

**Review by bkchr — APPROVED** [review]

File: Cargo.toml (hunk 12 above)
Comment (lexnv) [reply Cargo.toml:849]: We'll need another release for it once we merge: https://github.com/paritytech/frame-metadata/pull/86, sorry for the back and forth, wanted to add `transaction_extensions_by_version` as well 🙏

**Review by jsdw — APPROVED** [review]

**Review by gui1117 — APPROVED** [review]
Looks good to me, once we can use stable dependencies in the Cargo.toml

**Comment by lexnv** [comment]
/cmd prdoc --audience node_dev --bump patch --force

File: .github/workflows/check-semver.yml [hunk 14]
```diff
@@ -74,7 +74,7 @@ jobs:
 
       - name: install parity-publish
         # Set the target dir to cache the build.
-        run: CARGO_TARGET_DIR=./target/ cargo install parity-publish@0.8.0 --locked -q
+        run: CARGO_TARGET_DIR=./target/ cargo install parity-publish@0.10.0 --locked -q
```
Comment (lexnv) [inline .github/workflows/check-semver.yml:77]:
```suggestion
        run: CARGO_TARGET_DIR=./target/ cargo install parity-publish@0.10.1 --locked -q
```

**Review by pgherveou — APPROVED** [review]
for pallet-revive ownership

**Review by xermicus — APPROVED** [review]
